Add command line documentation for the create command (#1044)
terramate-io/terramate: docs/cmdline/create.md

---
title: terramate create - Command
description: With the terramate create command you can create a new stack in the current project.

> prev:

Author: will be updated later

---
title: terramate create - Command
description: With the terramate create command you can create a new stack in the current project.

prev:
  text: 'Stacks'
  link: '/stacks/'

next:
  text: 'Sharing Data'
  link: '/data-sharing/'
---

# Create

The `create` command creates a new stack. This command accepts a path that will
be used to create the stack in. The `create` command will create intermediate
directories as required.

## Usage

`terramate create [path] [options]`

## Examples

Create a new stack:

```bash
terramate create path/to/stack`
```

Complete Example:

```bash
terramate create path/to/stack \
    --name foo-stack \
    --description "This is an example" \
    --tags app,prd \
    --after ../../after-another-stack,../../after-this-stack \
    --before ../../before-this-stack,../../before-another-stack \
    --ignore-existing \
    --no-generate
```

## Options

> - `--id=STRING` ID of the stack. Defaults to UUID.

Reviewer: ```suggestion
- `--id=STRING` ID of the stack. Defaults to a random UUIDv4 (Using the default is highly recommended).
```

---
title: terramate create - Command
description: With the terramate create command you can create a new stack in the current project.

prev:
  text: 'Stacks'
  link: '/stacks/'

next:
  text: 'Sharing Data'
  link: '/data-sharing/'
---

# Create

The `create` command creates a new stack. This command accepts a path that will
be used to create the stack in. The `create` command will create intermediate
directories as required.

## Usage

Usage:

> `terramate create [path] [options]`

Author: @mariux is this correct or shall we rather say `terramate create [options] [path]` (which is what Terraform is using)

Reviewer: i would even say:

```suggestion
`terramate create [options] PATH`
```

as PATH is not optional bu required.

## Examples

Create a new stack:

```bash
terramate create path/to/stack`
```

Complete Example:

```bash
terramate create path/to/stack \
    --name foo-stack \
    --description "This is an example" \
    --tags "app" \
    --tags "prod" \
    --after ../../after-another-stack,../../after-this-stack \
    --before ../../before-this-stack \
    --before ../../before-another-stack \
    --ignore-existing \
    --no-generate
```

## Options

- `--id=STRING` ID of the stack. Defaults to UUID.
- `--name=STRING` Name of the stack. Defaults to stack dir base name.
- `--description=STRING` Description of the stack. Defaults to the stack name.

> - `--tags=STRING` Adds tags to the stack. `--tags A --tags B` is the same as `--tags A,B`.

Reviewer: ```suggestion
- `--tags=LIST` Adds tags to the stack. `--tags A --tags B` is the same as `--tags A,B`.
```

> - `--import=STRING`  Add import block for the given path on the stack.

Reviewer: ```suggestion
- `--import=LIST`  Add import block for the given path on the stack.
```

> - `--tags=STRING` Adds tags to the stack. `--tags A --tags B` is the same as `--tags A,B`.
> - `--import=STRING`  Add import block for the given path on the stack.
> - `--after=STRING` Add a stack as [after](../orchestration/index.md#explicit-order-of-execution). `--after ../after-this-stack --after after-another-stack` is the same as `--after ../after-this-stack,../after-another-stack`
> - `--before=STRING` Add a stack as [before](../orchestration/index.md#explicit-order-of-execution). `--before ../before-this-stack --before before-another-stack` is the same as `--before ../before-this-stack,../before-another-stack`

Reviewer: ```suggestion
- `--tags=LIST` Adds tags to the stack. `--tags A --tags B` is the same as `--tags A,B`.
- `--import=LIST`  Add import block for the given path on the stack.
- `--after=LIST` Add a stack as [after](../orchestration/index.md#explicit-order-of-execution). `--after ../after-this-stack --after after-another-stack` is the same as `--after ../after-this-stack,../after-another-stack`
- `--before=LIST` Add a stack as [before](../orchestration/index.md#explicit-order-of-execution). `--before ../before-this-stack --before before-another-stack` is the same as `--before ../before-this-stack,../before-another-stack`
```

Author: In that case this is unfortunately also documented wrong in the output of `--help`

Author: @mariux but this could be `LIST` or `STRING`

> - `--after=STRING` Add a stack as [after](../orchestration/index.md#explicit-order-of-execution). `--after ../after-this-stack --after after-another-stack` is the same as `--after ../after-this-stack,../after-another-stack`
> - `--before=STRING` Add a stack as [before](../orchestration/index.md#explicit-order-of-execution). `--before ../before-this-stack --before before-another-stack` is the same as `--before ../before-this-stack,../before-another-stack`

Reviewer: ```suggestion
- `--after=LIST`, `--before=LIST` Define an explicit [order of execution](../orchestration/index.md#explicit-order-of-execution). LIST can contain relative paths `../path/to/stack` and/or tags `tag:my-tag`. 
```

prev:
  text: 'Stacks'
  link: '/stacks/'

next:
  text: 'Sharing Data'
  link: '/data-sharing/'
---

# Create

The `create` command creates a new stack. This command accepts a path that will
be used to create the stack. The `create` command will create intermediate
directories as required. Whenever the `create` command is used, the code generation will run automatically to ensure all code relevant for the newly created stack will be generated.

## Usage

`terramate create [options] PATH`

## Examples

Create a new stack:

```bash
terramate create path/to/stack
```

Complete Example:

```bash
terramate create path/to/stack \
    --name foo-stack \
    --description "This is an example" \
    --tags app,prd \
    --after ../../after-another-stack,../../after-this-stack \
    --before ../../before-this-stack,../../before-another-stack \
    --ignore-existing \
    --no-generate
```

Initialize Terramate in an existing Terraform project:

```bash
terramate create --all-terraform
```

This is helpful when you want to onboard Terramate in an existing
Terraform project. `--all-terraform` will create a Terramate configuration

> file in every Terraform directory that contains a `terraform.backend` block.

Reviewer: ```suggestion
file in every Terraform directory that contain a `terraform.backend` block or `provider` blocks.
```

## Options

- `--id=STRING` ID of the stack. Defaults to a random UUIDv4 (Using the default is highly recommended).
- `--name=STRING` Name of the stack. Defaults to stack dir base name.
- `--description=STRING` Description of the stack. Defaults to the stack name.

> - `--tags=STRING` Adds tags to the stack. `--tags b --tags b` is the same as `--tags b,b`.

Reviewer: ```suggestion
- `--tags=STRING` Adds tags to the stack. Example: `--tags a --tags b.
```

there currently is a bug preventing `--tags a,b` also the text duplicates `b,b` instead of `a,b` ;) we can update the docs, once the bug is fixed.
